nav2_simple_commander/nav2_simple_commander/line_iterator.py

```diff
@@ -0,0 +1,201 @@
+#! /usr/bin/env python3
+# Copyright 2021 Samsung Research America
+# Copyright 2022 Afif Swaidan
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This is a Python3 API for a line iterator.
+
+It provides the ability to iterate through the points of a line.
+"""
+
+
+from cmath import sqrt
+
+
+class LineIterator():
+    def __init__(self, x0, y0, x1, y1, step_size=1.0):
+        """Initializer for LineIterator.
+
+        Args:
+            x0 (Float): Abscissa of the initial point
+            y0 (Float): Ordinate of the initial point
+            x1 (Float): Abscissa of the final point
+            y1 (Float): Ordinate of the final point
+            step_size (Float, optional): Resolution of increments. Defaults to 1.
+        """
+        self.x0_ = x0
+        self.y0_ = x0
+        self.x1_ = x1
+        self.y1_ = y1
+        self.x_ = x0
+        self.y_ = y0
+        self.step_size_ = step_size
+
+        if x1 != x0 and y1 != y0:
+            self.valid_ = True
+            self.m_ = (y1-y0)/(x1-x0)
+            self.b_ = y1 - (self.m_*x1)
+            if (self.b_ < 0):
+                self.equation_ = "y = " + str(self.m_) + "*x " + str(self.b_)
+            else:
+                self.equation_ = "y = " + str(self.m_) + "*x + " + str(self.b_)
+        elif x1 == x0 and y1 != y0:
+            self.valid_ = True
+            self.equation_ = "x = " + str(x1)
+        elif y1 == y1 and x1 != x0:
+            self.valid_ = True
+            self.m_ = (y1-y0)/(x1-x0)
+            self.b_ = y1 - (self.m_*x1)
+            self.equation_ = "y = " + str(y1)
+        else:
+            self.valid_ = False
+            self.equation_ = "Invalid"
+
+    def isValid(self):
+        """Returns True if the line is valid
+        Returns:
+            Bool: Flag if live is valid (true) or not (false)
+        """
+        return self.valid_
+
+    def advance(self):
```

Comment:
What about the very last point? If we have a terminus of `(x,y)`, we march towards it in `step_size` increments, but it might not _exactly_ hit `(x,y)`. Right now, we'd blow past it, then have `valid` be `false`, but never actual check the terminus point

Reply:
In the case of the very last point, if the step_size will increment in a way past the final coordinates (x,y) the "clamp" function that I'm using will make sure not to increment past the terminus of (x,y). This way we will always march on the line using step_size increments until the last increment that will be constrained not to cross the terminus of (x,y).

I tested it in a test script and it never crosses past the final point in the line while it increments according to step_size along the line.